docs/proposals/001-in-place-upgrades.md

```diff
@@ -0,0 +1,257 @@
+
+## Bootstrap Provider Changes
+<!--
+This section MUST mention any changes to the bootstrap provider.
+-->
+
+A machine controller called `MachineReconciler` is added which would perform the in-place upgrade if `k8sd.io/in-place-upgrade-to` annotation is set on the machine.
+
+The controller would use the value of this annotation to make an endpoint call to the `/x/capi/snap-refresh` through `k8sd-proxy`. 
+
+The result of this operation will be communicated back to the user via the `k8sd.io/in-place-upgrade-status` annotation. Values being:
+
+* `in-progress` for an upgrade currently in progress
+* `done` for a successful upgrade
+* `failed` for a failed upgrade
+
+After an upgrade process begins:
+* `k8sd.io/in-place-upgrade-status` annotation on the `Machine` would be added/updated with `in-progress`
+
+After a successfull upgrade:
+* `k8sd.io/in-place-upgrade-to` annotation on the `Machine` would be removed
+* `k8sd.io/in-place-upgrade-current` annotation on the `Machine` would be added/updated with the used `{upgrade-option}`.
+* `k8sd.io/in-place-upgrade-status` annotation on the `Machine` would be added/updated with `done`
+
+After a failed upgrade:
+* `k8sd.io/in-place-upgrade-failure` annotation on the `Machine` would be added/updated with the failure message
+* `k8sd.io/in-place-upgrade-status` annotation on the `Machine` would be added/updated with `failed`
+
+The reconciler should ignore the upgrade if `k8sd.io/in-place-upgrade-status` is already set to `in-progress` on the machine. 
+
+#### Changes for Rolling Upgrades and Creating New Machines
+In case of a rolling upgrade or when creating new machines the `CK8sConfigReconciler` should check for the `k8sd.io/in-place-upgrade-current` annotation both on the `Machine` and on the owner `Cluster` object.
+
+The value of one of these annotations should be used instead of the `version` field while generating a cloud-init script for a machine. The precedence of version fields are:
+1. Annotation on the `Machine`
+2. Annotation on the `Cluster`
+3. The `version` field
+
+Which means the value from the annotation on the `Machine` would be used first if found.
+
+Using an annotation value requires changing the `install.sh` file to perform the relevant snap operation based on the option.
+* `snap install k8s --classic --channel <channel>` for `Channel`
+* `snap install k8s --classic --revision <revision>` for `Revision`
+* `snap install <path-to-snap> --classic --dangerous --name k8s` for `LocalPath`
+
+When a rolling upgrade is triggered the `LocalPath` option requires the newly created machine to contain the local `*.snap` file. This usually means the machine image used by the infrastructure provider should be updated to contain this image. This file could possibly be sideloaded in the cloud-init script before installation.
+
+This operation should not be performed if `install.sh` script is overridden by the user in the manifests.
```

Comment:
How do we detect that?

Reply:
This will be done my making sure we don't override the `install.sh` script with the in-place options if there is a user provided file for it. We might not need to do anything specific other than making sure the `ExtraFiles` take precedence.

Reply:
We fully control the generated cloud-init here. So we could just replace the default install script with one that does whatever is needed based on the annotation (instead of using the embedded `scriptInstall`).

Then, if users have specified a custom `install.sh` script, the implementation already covers this case for us (it will come later in `extraFiles` of cloud-init, so it will override the default one.